### PR TITLE
Channel Packer improvements

### LINKED ISSUE #308: Improve channel packer
### Description

Here are some features that could improve the channel packer tool:

* Display image size at bottom of each texture
* Add a convert smoothness to roughness checkbox under Roughness, which inverts it
* Above Pack button, add checkboxes for:
    * Mipmap generation (default on) - Import template: mipmaps/generate=true
    High quality (default off) - Import template: compress/high_quality=true
* Enable and fix reading DDS files
* Resize 
* Terrain detiling fixes from Xtarsia

@epitaque

### LINKED ISSUE #467: Option to use noise in a height blending algorithm instead of a texture's heightmap
### Description

The height blending algorithm make overlay textures blend in much more naturally, but they rely on the texture's heightmap to support this behaviour.

heightmap_no_height_blending
*A patch of grass without height blending*

heightmap_height_blending
*Same patch of grass with height blending*

If there are no packed heightmaps in textures, this feature does practically nothing, aside from lowering or raising the opacity of the overlay:

no_heightmap_no_height_blending
*A patch of grass without heightmaps, no height blending*

no_heightmap_height_blending
*Same patch of grass, height blending*

Here is the code responsible for the feature (if I understand it correctly):

*function*
```glsl
vec4 height_blend(vec4 a_value, float a_height, vec4 b_value, float b_height, float blend) {
	if(height_blending) {
		float ma = max(a_height + (1.0 - blend), b_height + blend) - (1.001 - blend_sharpness);
	    float b1 = max(a_height + (1.0 - blend) - ma, 0.0);
	    float b2 = max(b_height + blend - ma, 0.0);
	    return (a_value * b1 + b_value * b2) / (b1 + b2);
	} else {
		float contrast = 1.0 - blend_sharpness;
		float factor = (blend - contrast) / contrast;
		return mix(a_value, b_value, clamp(factor, 0.0, 1.0));
	}
}
```

*calls to the function*
```glsl
void get_material(vec2 base_uv, uint control, ivec3 iuv_center, vec3 normal, out Material out_mat) {
    // other stuff...

    // Blend overlay and base
    albedo_ht = height_blend(albedo_ht, albedo_ht.a, albedo_ht2, albedo_ht2.a, out_mat.blend);
    normal_rg = height_blend(normal_rg, albedo_ht.a, normal_rg2, albedo_ht2.a, out_mat.blend);
    
    // other stuff...
}
```

## The proposal

Instead of just adding an additional modifier to overlay's blend vlaue when `height_blending` is checked off, we could instead use a noise-based blending:

*function*
```glsl
vec4 height_blend(vec4 a_value, float a_height, vec4 b_value, float b_height, float blend) {
    float ma = max(a_height + (1.0 - blend), b_height + blend) - (1.001 - blend_sharpness);
    float b1 = max(a_height + (1.0 - blend) - ma, 0.0);
    float b2 = max(b_height + blend - ma, 0.0);
    return (a_value * b1 + b_value * b2) / (b1 + b2);
}
```

*calls to the function*
```glsl
void get_material(vec2 base_uv, uint control, ivec3 iuv_center, vec3 normal, out Material out_mat) {
    // other stuff...

    float a_height;
    // Blend overlay and base
    if (height_blending) {
	a_height = albedo_ht.a;
    } else {
	a_height = texture(noise_texture, base_uv*noise3_scale).r;
    }
    albedo_ht = height_blend(albedo_ht, a_height, albedo_ht2, albedo_ht2.a, out_mat.blend);
    normal_rg = height_blend(normal_rg, a_height, normal_rg2, albedo_ht2.a, out_mat.blend);
    
    // other stuff...
}
```

heightmap_noise_blending
*A patch of grass with heightmaps, noise-based blending*

no_heightmap_noise_blending
*A patch of grass without heightmaps, noise-based blending*

This way there would be two approaches to choose from, and a better-functioning feature when not using textures packed with heightmaps. If the previous functionality is important, it can still be partially achieved with using textures without heighmaps and enabling height_blending (since the height will be the same at every point in the texel)

### PR DESCRIPTION
**Pull request by Xtarsia**
Closes #308 
Closes #467

Mipmap and Quality options
Luminance as Height
Invert Smoothness
Image Resizing
DDS file type support
normal alignment fix

Edit: now includes:
Inverted Depth
width and height display
selectable source channels for height / roughness
Drag n drop, including from OS

Edit2:
Fixed window management with defered dialogue calls and always ontop
some minor formatting
some code restructing
Luminance now a button that loads the generated height texture
Switched to Refcounted instead of Object as script base class.

image

### CODE REVIEW

**Comment by Xtarsia** [comment]
added width and height to each texture.

when resize is enabled, textures provided for packing no longer need to be the same size.

**Comment by Xtarsia** [comment]
This is ready for review!

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 1]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
 		save_file_dialog.title = "Save Packed Normal/Roughness Texture"
 		save_file_dialog.popup_centered_ratio()
 
-
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:179]: Two spaces between functions is correct.

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 2]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+		error = _pack_textures(images[IMAGE_ALBEDO], images[IMAGE_HEIGHT], p_dst_path, false,
+		invert_height_checkbox.button_pressed, false, height_channel_selected)
 	else:
-		_pack_textures(images[IMAGE_NORMAL], images[IMAGE_ROUGHNESS], p_dst_path, invert_green_checkbox.button_pressed)
+		error = _pack_textures(images[IMAGE_NORMAL], images[IMAGE_ROUGHNESS], p_dst_path,
+			invert_green_checkbox.button_pressed, invert_smooth_checkbox.button_pressed,
+			align_normals_checkbox.button_pressed, roughness_channel_selected)
+	
+	if error == OK:
+		plugin.get_editor_interface().get_resource_filesystem().scan()
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:333]: EditorInterface

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 3]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+	
+	return Basis(Vector3(v.x * v.x * k + c, vxy - v.z, vxz + v.y),
+		Vector3(vxy + v.z, v.y * v.y * k + c, vyz - v.x),
+		Vector3(vxz - v.y, vyz + v.x, v.z * v.z * k + c)
+	)
+
+func _set_normal_vector(source: Image, quiet: bool = false) -> void:
+	# Calculate texture normal sum direction
+	var normal:Image = source
+	var sum:Color = Color(0,0,0,0)
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:368]: Spaces after : and , on both lines, maybe decimal points too

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 4]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+	for x in normal.get_height():
+		for y in normal.get_width():
+			sum += normal.get_pixel(x, y)
+	var div:float = normal.get_height() * normal.get_width()
+	sum /= Color(div, div, div)
+	sum *= 2.0
+	sum -= Color(1,1,1)
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:375]: spaces and decimals

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 5]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+	normal_vector = Vector3(sum.r,sum.g,sum.b).normalized()
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:376]: spaces

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 6]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+	if normal_vector.dot(Vector3(0.0, 0.0, 1.0)) < 0.999 && !quiet:
+		_show_warn("Normal Texture Not Orthoganol to UV plane.\nFor Compatability with Detiling and Rotation, Select Orthoganolize Normals")
+
+func _align_normals(source: Image, iteration: int = 0) -> void:
+	# generate matrix to re-align the normalmap
+	var mat3:Basis = _alignment_basis(normal_vector)
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:382]: space :

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 7]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+			_show_success("Alignment OK, skipping Normal Orthogonalization")
+	
+		var output_image: Image = Terrain3DUtil.pack_image(p_rgb_image, p_a_image,
+			p_invert_green, p_invert_smooth, channel_to_pack)
+	
 		if not output_image:
 			_show_error("Failed to pack textures.")
-			return
-
+			return 1
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:423]: return FAILED

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 8]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+	
 		output_image.save_png(p_dst_path)
-		editor_interface.get_resource_filesystem().scan_sources()
 		_create_import_file(p_dst_path)
 		_show_success("Packed to " + p_dst_path + ".")
+		return 0
 	else:
 		_show_error("Failed to load one or more textures.")
+		return 1
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:431]: Return OK and Failed

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 9]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
 	if p_rgb_image and p_a_image:
-		if p_rgb_image.get_size() != p_a_image.get_size():
-			_show_error("Textures must be the same size.")
-			return
-
-		var output_image: Image = Terrain3DUtil.pack_image(p_rgb_image, p_a_image, p_invert_green)
-
+		if p_rgb_image.get_size() != p_a_image.get_size() and !resize_toggle_checkbox.button_pressed:
+			_show_error("Textures must be the same size.\nEnable resize to override image dimensions")
+			return 1
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:406]: Return FAILED

File: project/addons/terrain_3d/src/channel_packer.gd [hunk 10]
```diff
@@ -176,37 +318,114 @@ func _on_pack_button_pressed() -> void:
...
+			vector_pixel *= 0.5
+			var new_pixel:Color = Color(vector_pixel.x, vector_pixel.y, vector_pixel.z, old_pixel.a)
+			source.set_pixel(x, y, new_pixel)
+	_set_normal_vector(source, true)
+	if normal_vector.dot(Vector3(0.0, 0.0, 1.0)) < 0.999 && iteration < 3:
+		++iteration
+		_align_normals(source, iteration)
+
+func _pack_textures(p_rgb_image: Image, p_a_image: Image, p_dst_path: String, p_invert_green: bool,
+	p_invert_smooth: bool, p_align_normals : bool, channel_to_pack: int) -> int:
```
Comment (TokisanGames) [inline project/addons/terrain_3d/src/channel_packer.gd:402]: Return type Error

**Review by TokisanGames — CHANGES_REQUESTED** [review]
Great job. Thanks for improving this. Here are some notes:

* Sometimes generating the height texture and packing produces DXT1, no alpha. This happened after selecting a texture for the height slot, then generating a height texture. In fact, closing the window doesn't reset it. I have to restart the project and generate to get any texture with alpha.

* Please use _gui_input() to close the popup on Escape. Making it a real popup would include that, but I think you did different settings.

* The generated height is better than the artist provided height sometimes. Which tells me we need to improve the blending options and give a height adjustment in a future PR. #351

**Comment by Xtarsia** [comment]
Rebased, all reviewed points addressed.

window now closes on ESC.

The error im looking into a bit more.

**Comment by TokisanGames** [comment]
Great work, thank you!

**Pull request merged**